pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -545,6 +545,17 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if err != nil {
 			return err
 		}
+		if !found {
+			// If PV is pre-bound by scheduler, it's possible on heavy load
+			// that corresponding PVC is not synced in controller yet. So we
+			// double-check PVC in apiserver to make sure we will not reclaim a
+			// PV due to API delay.
+			obj, err = ctrl.kubeClient.CoreV1().PersistentVolumeClaims(volume.Spec.ClaimRef.Namespace).Get(volume.Spec.ClaimRef.Name, metav1.GetOptions{})
```

Comment:
Is it better to check the storage-class's volumeBindingMode=WaitForFirstConsumer before get the pvc from api-server?

Reply:
It should be ok to check PVC in apiserver only if storage class volumeBindingMode is `WaitForFirstConsumer` normally, but the cause of problem is API delay, checking class here add some complexity.
@msau42 @jsafrane
What do you think?

Reply:
One scenario is if there are lots of PVCs are deleted, then there will be many requests send to api server.